src/xb-node.c

```diff
@@ -258,6 +284,76 @@ xb_node_get_children (XbNode *self)
 	return array;
 }
 
+/**
+ * xb_node_child_iter_init:
+ * @iter: an uninitialized #XbNodeChildIter
+ * @self: a #XbNode
+ *
+ * Initializes a child iterator for the node's children and associates
+ * it with @self.
+ * The #XbNodeChildIter structure is typically allocated on the stack
+ * and does not need to be freed explicitly.
+ *
+ * Since: 0.3.4
+ */
+void
+xb_node_child_iter_init (XbNodeChildIter *iter, XbNode *self)
+{
+	XbNodePrivate *priv = GET_PRIVATE (self);
+	RealChildIter *ri = (RealChildIter *) iter;
+
+	g_return_if_fail (iter != NULL);
+	g_return_if_fail (XB_IS_NODE (self));
+
+	ri->node = self;
+	ri->position = xb_silo_get_child_node (priv->silo, priv->sn);
+}
+
+/**
+ * xb_node_child_iter_next:
+ * @iter: an initialized #XbNodeAttrIter
+ * @child: (out) (optional) (not nullable): Destination of the returned child
+ *
+ * Returns the current child and advances the iterator.
+ * The retrieved #XbNode child needs to be dereferenced with g_object_unref().
+ * Example:
+ * |[<!-- language="C" -->
+ * XbNodeChildIter iter;
+ * g_autoptr(XbNode) child = NULL;
+ *
+ * xb_node_child_iter_init (&iter, node);
+ * while (xb_node_child_iter_next (&iter, &child)) {
+ *     // do something with the node child
+ *     g_clear_pointer (&child, g_object_unref);
```

Comment:
You might want to also consider adding something similar to `g_variant_iter_loop()`. It’s a version of `g_variant_iter_next()` which handles the memory management of `child` for you, which saves making users remember to add this `g_clear_pointer()` call.

It’s not usable from languages other than C, though, which is why `g_variant_iter_next()` still exists. You might want to consider a similar arrangement here.

Reply:
Interesting! I think I would need to carry an extra bool around with that then, to know the first invocation of that loop function...